File: app/src/main/java/com/vdovin/spacex/screen/main/structure/LaunchesPresenter.java
package com.vdovin.spacex.screen.main.structure;

import com.vdovin.spacex.base.BasePresenter;
import com.vdovin.spacex.database.dao.SpaceXDao;
import com.vdovin.spacex.rx.RxSchedulers;

import io.reactivex.disposables.Disposable;

public class LaunchesPresenter extends BasePresenter<LaunchesView> {

    private SpaceXDao spaceXDao;
    private RxSchedulers rxSchedulers;

    public LaunchesPresenter(SpaceXDao spaceXDao, RxSchedulers rxSchedulers) {
        this.spaceXDao = spaceXDao;
        this.rxSchedulers = rxSchedulers;
    }

    @Override
    protected void onLoad() {
        super.onLoad();
        disposables.add(loadAllLaunches());
        disposables.add(launchClicked());
    }

    private Disposable loadAllLaunches() {
        return spaceXDao.getAllLaunches()
                .subscribeOn(rxSchedulers.io())
                .observeOn(rxSchedulers.androidUI())
                .subscribe(launches -> getView().displayLaunches(launches));
    }

    private Disposable launchClicked() {
        return getView().launchClicked().subscribe(spaceX -> getView().openDetails(spaceX));
    }

}

File: app/src/main/java/com/vdovin/spacex/screen/main/LaunchesPresenter.kt
package com.vdovin.spacex.screen.main

import com.vdovin.spacex.base.BasePresenter
import com.vdovin.spacex.database.dao.SpaceXDao
import kotlinx.coroutines.withContext

class LaunchesPresenter(private val dao: SpaceXDao) : BasePresenter<LaunchesView>() {
    override fun start() {
        launch {
            val launches = withContext(backgroundContext) { dao.allLaunches }
            view.displayLaunches(launches)
            view.launchClicked{ spaceX -> view.openDetails(spaceX) }
        }
    }
}